pkg/multus/multus.go

```diff
@@ -583,50 +606,148 @@ func CmdAdd(args *skel.CmdArgs, exec invoke.Exec, kubeClient *k8s.ClientInfo) (c
 		return nil, cmdErr(nil, "error getting k8s args: %v", err)
 	}
 
-	if n.ReadinessIndicatorFile != "" {
+	pod, err := getPod(kubeClient, k8sArgs, false)
+	if err != nil {
+		return nil, err
+	}
+
+	netnsPath := args.Netns
+	if netns, found := os.LookupEnv("CNI_NETNS"); found {
+		netnsPath = netns
+	}
+
+	ifName := args.IfName
+	if interfaceName, found := os.LookupEnv("CNI_IFNAME"); found {
+		ifName = interfaceName
+	}
+	result, err := NewMultusCmd(
+		args.ContainerID,
+		string(k8sArgs.K8S_POD_INFRA_CONTAINER_ID),
+		ifName,
+		netnsPath,
+		string(k8sArgs.K8S_POD_NAME),
+		string(k8sArgs.K8S_POD_NAMESPACE),
+		string(k8sArgs.K8S_POD_UID)).CmdAdd(n, pod, exec, kubeClient)
+	if err != nil {
+		return nil, fmt.Errorf("%v", err)
+	}
+	return result, nil
+}
+
+//CmdCheck ...
+func CmdCheck(args *skel.CmdArgs, exec invoke.Exec, kubeClient *k8s.ClientInfo) error {
+	in, err := types.LoadNetConf(args.StdinData)
+	logging.Debugf("CmdCheck: %v, %v, %v", args, exec, kubeClient)
+	if err != nil {
+		return err
+	}
+
+	k8sArgs, err := k8s.GetK8sArgs(args)
+	if err != nil {
+		return cmdErr(nil, "error getting k8s args: %v", err)
+	}
+
+	return NewMultusCmd(
+		args.ContainerID,
+		string(k8sArgs.K8S_POD_INFRA_CONTAINER_ID),
+		args.IfName,
+		args.Netns,
+		string(k8sArgs.K8S_POD_NAME),
+		string(k8sArgs.K8S_POD_NAMESPACE),
+		string(k8sArgs.K8S_POD_UID)).CmdCheck(in, exec)
+}
+
+//CmdDel ...
+func CmdDel(args *skel.CmdArgs, exec invoke.Exec, kubeClient *k8s.ClientInfo) error {
+	in, err := types.LoadNetConf(args.StdinData)
+	logging.Debugf("CmdDel: %v, %v, %v", args, exec, kubeClient)
+	if err != nil {
+		return err
+	}
+
+	k8sArgs, err := k8s.GetK8sArgs(args)
+	if err != nil {
+		return cmdErr(nil, "error getting k8s args: %v", err)
+	}
+
+	kubeClient, err = k8s.GetK8sClient(in.Kubeconfig, kubeClient)
+	if err != nil {
+		return cmdErr(nil, "error getting k8s client: %v", err)
+	}
+
+	pod, err := getPod(kubeClient, k8sArgs, true)
+	if err != nil {
+		return err
+	}
+
+	return NewMultusCmd(
+		args.ContainerID,
+		string(k8sArgs.K8S_POD_INFRA_CONTAINER_ID),
+		args.IfName,
+		args.Netns,
+		string(k8sArgs.K8S_POD_NAME),
+		string(k8sArgs.K8S_POD_NAMESPACE),
+		string(k8sArgs.K8S_POD_UID),
+	).CmdDelete(in, pod, exec, kubeClient)
+}
+
+// NewMultusCmd return a new ADD / CHECK / DEL command
+func NewMultusCmd(containerID, sandboxID, ifaceName, netnsPath, podName, podNamespace, podUID string) *CNICmd {
+	return &CNICmd{
+		ContainerID:   containerID,
+		SandboxID:     sandboxID,
+		InterfaceName: ifaceName,
+		NetnsPath:     netnsPath,
+		PodName:       podName,
+		PodNamespace:  podNamespace,
+		PodUID:        podUID,
+	}
+}
+
+// CmdAdd is the handler of the CNI ADD command
+func (mc *CNICmd) CmdAdd(cniConf *types.NetConf, pod *v1.Pod, exec invoke.Exec, kubeclient *k8s.ClientInfo) (cnitypes.Result, error) {
```

Comment:
To me, `mc *CNICmd` seems to be part of args, not method variable.

Reply:
I don't follow this comment. I'm making `mc` the receiver for this function, which makes it a method.

The overall idea is for both the "old" binary (current one) and the new multus daemon should create a `CNICmd` with the required immutable data (args / cmd / netns / container id / ...) and just call the respective methods.

What do you mean w/ "be part of args" ?

Reply:
Before the PR, it is function. Now it is method, even though you can change it as following:

```
func CmdAdd(mc *CNICmd, cniConf *types.NetConf, pod *v1.Pod, exec invoke.Exec, kubeclient *k8s.ClientInfo) (cnitypes.Result, error) {
```